File: exercises/exercises-mpi-p/viewwrite.py
#!/usr/bin/env python
################################################################
################################################################
####
#### This program file is part of the book and course
####   Parallel Computing for Science and Engineering
#### by Victor Eijkhout, copyright 2013-2020
####
#### viewwrite.py : MPI python exercise for MPI I/O
####
################################################################
################################################################

from mpi4py import MPI
import numpy as np
import struct
import time

comm = MPI.COMM_WORLD

nprocs = comm.Get_size()
procno = comm.Get_rank()

##
## Each process has a buffer of length `nwords'
## containing consecutive integers
##
nwords = 3
output_data = np.empty(nwords,dtype=np.intc)
for iw in range(nwords):
    output_data[iw] = procno*nwords + iw + 1


##
## Create the output file and open for write
##
mpifile = MPI.File.Open\
          (comm,"blockwrite.dat",
           amode=MPI.MODE_CREATE+MPI.MODE_WRONLY)


##
## Exercise:
## -- set a file view so that each processor writes at a different offset
## -- note: offsets are in bytes
##
sizeofint = np.dtype('intc').itemsize
mpifile.Set_view(
#### your code here ####
    MPI.INT)

nwriters = 2
if procno<nwriters:
    mpifile.Write( [output_data,nwords,MPI.INT] )

mpifile.Close()

if procno==0:
    f = open("blockwrite.dat","rb")
    errors = 0
    for i in range(nwords*nwriters):
        bytestring = f.read(sizeofint)
        ## python 3:
        #### fromfile = int.from_bytes( bytestring )
        ## python 2
        fromfile = struct.unpack("<L",bytestring)[0]
        if fromfile!=i+1:
            print("Error: s/b %d, got %d" % (i+1,fromfile))
            errors += 1
    f.close()
    if errors==0:
        print("Success. File correctly written.")
    else:
        print("Failed with %d errors" % errors)

File: exercises/exercises-mpi-cxx/viewwrite.cxx
/* %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%
   %%%% This program file is part of the book and course
   %%%%   "Parallel Computing for Science and Engineering"
   %%%% by Victor Eijkhout, copyright 2013-9
   %%%%
   %%%% viewwrite.c : MPI IO example
   %%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
*/

#include <iostream>
#include <sstream>
using namespace std;
#include <mpi.h>

int main(int argc,char **argv) {

  MPI_Comm comm;
  int nprocs,procno;
  stringstream proctext;

  MPI_Init(&argc,&argv);
  comm = MPI_COMM_WORLD;
  MPI_Comm_size(comm,&nprocs);
  MPI_Comm_rank(comm,&procno);

  /*
   * Each process has a buffer of length `nwords'
   * containing consecutive integers
   */
  int nwords = 3;
  int *output_data = (int*) malloc(nwords*sizeof(int));
  for (int iw=0; iw<nwords; iw++)
    output_data[iw] = procno*nwords+iw+1;
  
  /*
   * Create the output file and open for write
   */
  MPI_File mpifile;
  MPI_File_open(comm,"blockwrite.dat",
		MPI_MODE_CREATE | MPI_MODE_WRONLY,MPI_INFO_NULL,
		&mpifile);

  /*
   * Exercise:
   * -- set a file view so that each processor writes at a different offset
   * -- Note: offsets are in bytes
   */
  MPI_File_set_view
    (mpifile,
/**** your code here ****/
     MPI_INT,"native",MPI_INFO_NULL);

  int nwriters = 2;
  if (procno<nwriters) {
    MPI_File_write // no offset, because that is taken care of by the view
      (mpifile,output_data,nwords,MPI_INT,MPI_STATUS_IGNORE);
  }

  MPI_File_close(&mpifile);

  if (procno==0) {
    int error=nprocs, errors;
    FILE *f;
    f = fopen("blockwrite.dat","r");
    for (int i=0; i<nwords*nwriters; i++) {
      int fromfile,success;
      success = fread(&fromfile,sizeof(int),1,f);
      if (success==EOF) {
	error = procno;
	proctext << "Premature end of file";
	cerr << proctext.str() << "\n";
	break;
      }
      if (fromfile!=i+1) {
	error = procno;
  	proctext << "Error s/b " << i+1 << ", got " << fromfile;
	cerr << proctext.str() << "\n";
      }
    }
    fclose(f);
    errors = error;
    if (errors==nprocs) 
      proctext << "Execution finished correctly";
    else
      proctext << "Execution finished with errors";
    cerr << proctext.str() << "\n"; 
  }

  MPI_Finalize();
  return 0;
}
